libp2p/builders.nim

```diff
@@ -0,0 +1,206 @@
+import
+  options, tables, chronos, bearssl,
+  switch, peerid, peerinfo, stream/connection, multiaddress,
+  crypto/crypto, transports/[transport, tcptransport],
+  muxers/[muxer, mplex/mplex],
+  protocols/[identify, secure/secure, secure/noise],
+  connmanager
+
+export
+  switch, peerid, peerinfo, connection, multiaddress, crypto
+
+type
+  SecureProtocol* {.pure.} = enum
+    Noise,
+    Secio {.deprecated.}
+
+  SwitchBuilder* {.byref.} = object
+    bprivKey: Option[PrivateKey]
+    baddress: MultiAddress
+    bsecureManagers: seq[SecureProtocol]
+    btransportFlags: set[ServerFlags]
+    brng: ref BrHmacDrbgContext
+    binTimeout: Duration
+    boutTimeout: Duration
+    bmaxConnections: int
+    bmaxIn: int
+    bmaxOut: int
+    bmaxConnsPerPeer: int
+    bprotoVersion: Option[string]
+    bagentVersion: Option[string]
+
+proc init*(_: type[SwitchBuilder]): SwitchBuilder =
+  SwitchBuilder(
+    bprivKey: none(PrivateKey),
+    baddress: MultiAddress.init("/ip4/127.0.0.1/tcp/0").tryGet(),
+    bsecureManagers: @[],
+    btransportFlags: {},
+    brng: newRng(),
+    binTimeout: 5.minutes,
+    boutTimeout: 5.minutes,
+    bmaxConnections: MaxConnections,
+    bmaxIn: -1,
+    bmaxOut: -1,
+    bmaxConnsPerPeer: MaxConnectionsPerPeer,
+    bprotoVersion: none(string),
+    bagentVersion: none(string))
+
+# so I tried using var inpit and return but did not work...
+# as in proc privateKey*(builder: var SwitchBuilder, privateKey: PrivateKey): var SwitchBuilder =
+# so in nim we are stuck with this hardly efficient way and hopey compiler figures it out.. heh
+# maybe {.byref.} works.... I would not bet on it but let's use it.
+
+proc privateKey*(builder: SwitchBuilder, privateKey: PrivateKey): SwitchBuilder =
+  var b = builder
+  b.bprivKey = some(privateKey)
+  b
+
+proc address*(builder: SwitchBuilder, address: MultiAddress): SwitchBuilder =
+  var b = builder
+  b.baddress = address
+  b
+
+proc secureManager*(builder: SwitchBuilder, secureManager: SecureProtocol): SwitchBuilder =
+  var b = builder
+  b.bsecureManagers &= secureManager
+  b
+
+proc transportFlag*(builder: SwitchBuilder, flag: ServerFlags): SwitchBuilder =
```

Comment:
Hmm, I think we need to differentiate which transport is being built here. I would probably use something like `withTcpTransport(flags = ServerFlags)`. Same for muxers and secure managers. The `with` is not necessary, but I've seen this convention used pretty widely, but I'm either way on it.

Reply:
Good point,
About the `with`, rust builders generally don't have any extra words.
Some random example I stumbled upon in the last months:
https://docs.rs/rapier3d/0.6.1/rapier3d/dynamics/struct.RigidBodyBuilder.html
https://docs.rs/web3/0.15.0/web3/contract/deploy/struct.Builder.html
That was my style rationale.

Reply:
> About the with, rust builders generally don't have any extra words.

I do confess that I like `with` better than just the bare word and nim tends to use camel case vs snake case in rust. But, like I said it's a small detail and I'm either way on it.